### PR TITLE
[FEATURE]Horovod support for training transformer (PART 2)

### PR DESCRIPTION
**Pull request by hymzoque**
### Essentials ###
- [x] PR's title starts with a category (e.g. [BUGFIX], [MODEL], [TUTORIAL], [FEATURE], [DOC], etc)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage
- [ ] Code is well-documented

### Changes ###
-  add horovodrun example in machine_translation/README.md 
-  transformer base/large result (test and valid bleu)
-  ShardedIterator (to split sampler) and testcase
-  add test case to test situation that fail downloading in test_util_misc.py

### CODE REVIEW

File: scripts/machine_translation/README.md [hunk 1]
```diff
@@ -30,30 +30,38 @@ python3 train_transformer.py \
...
+    --end 39 \
+    --save-path transformer_base_wmt2014_en_de_${SUBWORD_ALGO}/epoch_avg_30_39.params
 ```
 
-
 Use the following command to inference/evaluate the Transformer model:
 
 ```bash
 SUBWORD_MODEL=yttm
-python3 evaluate_transformer.py \
```
Comment (ZheyuYe) [inline scripts/machine_translation/README.md:55]: plz use python3 so that it works on any systems

File: scripts/datasets/machine_translation/wmt2014_ende.sh [hunk 2]
```diff
@@ -23,7 +23,7 @@ nlp_preprocess clean_tok_para_corpus --src-lang ${SRC} \
                       --src-corpus train.raw.${SRC} \
                       --tgt-corpus train.raw.${TGT} \
                       --min-num-words 1 \
-                      --max-num-words 100 \
+                      --max-num-words 250 \
```
Comment (sxjscience) [inline scripts/datasets/machine_translation/wmt2014_ende.sh:26]: I think 250 is the truncation length after BPE tokenization. Have you verified that we should truncate to 250 before the BPE tokenization?

File: scripts/machine_translation/README.md [hunk 3]
```diff
@@ -97,7 +105,7 @@ python3 train_transformer.py \
     --gpus 0,1,2,3
 ```
 
-Use the average_checkpoint cli to average the last 10 checkpoints
+Use the average_checkpoint cli to average the last 10 update checkpoints
```
Comment (sxjscience) [inline scripts/machine_translation/README.md:108]: Just use `average the last 10 checkpoints`.

File: tests/test_data_sampler.py [hunk 4]
```diff
@@ -137,13 +137,18 @@ def test_split_sampler_even_size(num_samples, num_parts):
...
+    samplers = []
+    for index in range(num_parts):
+        samplers.append(s.BoundedBudgetSampler(seq_lengths, max_num_tokens, max_num_sentences,
+                                               required_batch_size_multiple, shuffle, seed,
+                                               num_parts, index))
     total_sampled_ids = []
-    for batch_sample_ids in sampler:
-        total_sampled_ids.extend(batch_sample_ids)
+    for sampler in samplers:
+        print(sampler)
```
Comment (ZheyuYe) [inline tests/test_data_sampler.py:150]: Should we avoid printing in test?

File: scripts/machine_translation/train_transformer.py [hunk 5]
```diff
@@ -328,9 +333,9 @@ def train(args):
     cfg.freeze()
     model = TransformerModel.from_cfg(cfg)
     model.initialize(mx.init.Xavier(magnitude=args.magnitude),
-                     ctx=ctx_l)
+                     ctx=args.ctx_l)
```
Comment (sxjscience) [inline scripts/machine_translation/train_transformer.py:336]: I think you should use `ctx_l`.

File: scripts/machine_translation/train_transformer.py [hunk 6]
```diff
@@ -457,7 +461,7 @@ def train(args):
                 sample_data_l = next(train_multi_data_loader)
             except StopIteration:
                 is_last_batch = True
-            if local_rank == 0 and num_params is None:
+            if args.local_rank == 0 and num_params is None:
```
Comment (sxjscience) [inline scripts/machine_translation/train_transformer.py:464]: Why use args.local_rank instead of local_rank?

File: scripts/machine_translation/train_transformer.py [hunk 7]
```diff
@@ -475,35 +479,39 @@ def train(args):
...
                    (n_epoch_train_iters % args.log_interval == 0 or is_last_batch):
                     log_end_time = time.time()
                     log_wc = log_wc.asnumpy()
                     wps = log_wc / (log_end_time - log_start_time)
                     log_avg_loss = (log_avg_loss / log_loss_denom).asnumpy()
                     logging.info('[Epoch {} Batch {}/{}] loss={:.4f}, ppl={:.4f}, '
                                  'throughput={:.2f}K wps, wc={:.2f}K, LR={}'
-                                 .format(epoch_id, processed_batch_num * num_parts, len(train_data_loader),
-                                         log_avg_loss, np.exp(log_avg_loss),
+                                 .format(epoch_id, min(processed_batch_num * args.num_parts, len(train_data_loader)),
```
Comment (sxjscience) [inline scripts/machine_translation/train_transformer.py:490]: Why do we need the min?

File: src/gluonnlp/data/sampler.py [hunk 8]
```diff
@@ -289,12 +288,16 @@ class BoundedBudgetSampler(BaseSampler):
...
+        If the number of batches is not even across all partitions, sample a few extra batches
+        for the ones with fewer batches.
     """
     def __init__(self, lengths: Union[Sequence[int], Sequence[Sequence[int]]],
                  max_num_tokens: int = -1, max_num_sentences: int = -1,
                  required_batch_size_multiple: int = 1,
                  shuffle: bool = True, seed: Optional[int] = None,
-                 num_parts: int = 1, part_index: int = 0):
+                 num_parts: int = 1, part_index: int = 0,
+                 even_size: bool = False):
```
Comment (sxjscience) [inline src/gluonnlp/data/sampler.py:300]: I suggest wrapping the handling of `even_size` and `num_parts` into another function.

Comment (sxjscience) [reply src/gluonnlp/data/sampler.py:300]: Is it possible to write something similar to https://github.com/NVIDIA/DeepLearningExamples/blob/d875531dd69de92c8424e2254df6f393088e7ef4/PyTorch/Translation/Transformer/fairseq/data/data_utils.py#L65-L89?

File: scripts/datasets/machine_translation/wmt2014_ende.sh (hunk 2 above)
Comment (hymzoque) [reply scripts/datasets/machine_translation/wmt2014_ende.sh:26]: 250 is consistent with  https://github.com/NVIDIA/DeepLearningExamples/blob/master/PyTorch/Translation/Transformer/examples/translation/prepare-wmt14en2de.sh#L181-L182

File: scripts/machine_translation/train_transformer.py (hunk 7 above)
Comment (hymzoque) [reply scripts/machine_translation/train_transformer.py:490]: `processed_batch_num * args.num_parts > len(train_data_loader)` may happens

File: tests/test_data_sampler.py (hunk 4 above)
Comment (hymzoque) [reply tests/test_data_sampler.py:150]: idk, it's consistent with https://github.com/dmlc/gluon-nlp/blob/master/tests/test_data_sampler.py#L37

File: scripts/machine_translation/train_transformer.py (hunk 7 above)
Comment (sxjscience) [reply scripts/machine_translation/train_transformer.py:490]: I think for the purpose of logging, we should just use `processed_batch_num * args.num_parts`.

**Comment by sxjscience** [comment]
For the diff check failure, it might be caused by the fact that the exception is not triggered:
https://github.com/dmlc/gluon-nlp/blob/32e87d4d4aa20a6eb658ee90d765ccffbd160571/src/gluonnlp/utils/misc.py#L562-L568

Thus, we may try to add a test-case which downloads an non-existing file and use `pytest.assertRaises`.

File: tests/test_data_sampler.py (hunk 4 above)
Comment (sxjscience) [reply tests/test_data_sampler.py:150]: It's not necessary to `print(sampler)`. However, it does not introduce any bugs because the output stream will be ignored.

File: scripts/machine_translation/train_transformer.py (hunk 7 above)
Comment (hymzoque) [reply scripts/machine_translation/train_transformer.py:490]: ok

File: tests/test_data_sampler.py (hunk 4 above)
Comment (hymzoque) [reply tests/test_data_sampler.py:150]: ok

File: scripts/datasets/machine_translation/wmt2014_ende.sh (hunk 2 above)
Comment (hymzoque) [reply scripts/datasets/machine_translation/wmt2014_ende.sh:26]: it should be 100

File: scripts/machine_translation/README.md [hunk 9]
```diff
@@ -30,30 +30,38 @@ python3 train_transformer.py \
...
     --gpus 0,1,2,3
 ```
 
-Use the average_checkpoint cli to average the last 10 checkpoints
+Or training via horovod
+```
+mpirun -np 4 -H localhost:4 python3 train_transformer.py \
+    --comm_backend horovod \
+    ...
+```
```
Comment (sxjscience) [inline scripts/machine_translation/README.md:48]: Do you have any special reason to use mpirun instead of horovodrun? @ZheyuYe

Comment (ZheyuYe) [reply scripts/machine_translation/README.md:48]: At first I cant run the scripts via 'horovodrun' and switched to 'mpirun'，probably because I didn't install it properly. Let me re-try 'horovodrun'.

File: src/gluonnlp/data/sampler.py (hunk 8 above)
Comment (hymzoque) [reply src/gluonnlp/data/sampler.py]: changed to ShardedIterator class in sampler.py

File: scripts/machine_translation/README.md (hunk 9 above)
Comment (hymzoque) [reply scripts/machine_translation/README.md:48]: changed mpirun to horovodrun

**Comment by sxjscience** [comment]
I find that the codecov is not very stable, especially the diff hit. @szha @leezu Would you know how could we improve it?

**Comment by szha** [comment]
assuming no race issue in result upload, this can be a result of non-determinism in tests, and the code path in the same test can vary.

if that's the case, we can mock the randomness to deterministically test each path. or if we still require the randomness, we can increase the trial number

File: scripts/machine_translation/README.md [hunk 10]
```diff
@@ -30,30 +30,37 @@ python3 train_transformer.py \
...
     --warmup_init_lr 0.0 \
     --seed 123 \
     --gpus 0,1,2,3
 ```
 
+Or training via horovod
+```
+horovodrun -np 4 -H localhost:4 python3 train_transformer.py \
+    --comm_backend horovod \
+    ...
```
Comment (sxjscience) [inline scripts/machine_translation/README.md:47]: Rather than `...`, would you paste the whole command?
